rpc/json-rpc/src/response.rs

```diff
@@ -0,0 +1,402 @@
+    ///
+    /// ```rust
+    /// use json_rpc::{Id, Response, error::{ErrorObject, ErrorCode}};
+    ///
+    /// let ok = Response::<()>::invalid_params(Id::Num(0));
+    /// let json = serde_json::to_string(&ok).unwrap();
+    /// assert_eq!(json, r#"{"jsonrpc":"2.0","id":0,"error":{"code":-32602,"message":"Invalid params"}}"#);
+    /// ```
+    pub const fn invalid_params(id: Id) -> Self {
+        Self {
+            jsonrpc: Version,
+            payload: Err(ErrorObject::invalid_params()),
+            id,
+        }
+    }
+
+    /// Creates an error response using [`ErrorObject::internal_error`].
+    ///
+    /// ```rust
+    /// use json_rpc::{Id, Response, error::{ErrorObject, ErrorCode}};
+    ///
+    /// let ok = Response::<()>::internal_error(Id::Num(0));
+    /// let json = serde_json::to_string(&ok).unwrap();
+    /// assert_eq!(json, r#"{"jsonrpc":"2.0","id":0,"error":{"code":-32603,"message":"Internal error"}}"#);
+    /// ```
+    pub const fn internal_error(id: Id) -> Self {
+        Self {
+            jsonrpc: Version,
+            payload: Err(ErrorObject::internal_error()),
+            id,
+        }
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- Trait impl
+impl<T> std::fmt::Display for Response<T>
+where
+    T: Clone + Serialize,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match serde_json::to_string_pretty(self) {
+            Ok(json) => write!(f, "{json}"),
+            Err(_) => Err(std::fmt::Error),
+        }
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- Key
+/// This type represents the `"key"` values within [`Response`].
+///
+/// We need this since `Response` has a custom deserializer implementation.
+pub(crate) enum Key {
+    /// "jsonrpc" field.
+    JsonRpc,
+    /// "result" field.
+    Result,
+    /// "error" field.
+    Error,
+    /// "id" field.
+    Id,
+}
+
+/// Serde [`Key`] visitor marker struct.
+struct KeyVisitor;
+
+impl serde::de::Visitor<'_> for KeyVisitor {
+    type Value = Key;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("Key field must be a string and one of the following values: ['jsonrpc', 'result', 'error', 'id']")
+    }
+
+    fn visit_str<E: serde::de::Error>(self, text: &str) -> Result<Self::Value, E> {
+        // FIXME: I don't think JSON-RPC 2.0 specifies
+        // case-sensitiveness so just be strict here.
+        Ok(match text {
+            "jsonrpc" => Key::JsonRpc,
+            "result" => Key::Result,
+            "error" => Key::Error,
+            "id" => Key::Id,
+            _ => {
+                return Err(serde::de::Error::invalid_value(
+                    serde::de::Unexpected::Str(text),
+                    &self,
+                ))
+            }
+        })
```

Comment:
How should we treat case? I.e. should this work?:
```json
{"jSoNRPC":"2.0","iD":"0","method":"get_block"}
```
AFAICT `monerod` doesn't care about case except for both key and value of `method/params`, so the above works.

Reply:
Also worth noting if we support mixed-case then I think we have to do a manual deserialization impl for `Request` as well.

Reply:
I think we can get away with only accepting lower case

Reply:
It's also worth noting we error instead of ignoring extra fields, unlike `monerod` and other [JSON-RPC 2.0 impls](https://docs.rs/jsonrpsee-types/0.23.1/src/jsonrpsee_types/response.rs.html#198).

I.e.:
```json
{"hello":0,"jsonrpc":"2.0","id":"0","method":"get_block"}
```
This is OK for `monerod` but will error for us. I think we need to do a manual deserialization impl to support this.

Reply:
> This means to be compliant with no result and error we must use #[serde(deny_unknown_fields)], which means we will error on extra fields

Both `result` and `error` would go to the same field which I _think_ serde disallows, might be worth adding a test to make sure?

Reply:
I think the tag is needed, commenting it out will allow serializing to pass: https://github.com/Cuprate/cuprate/pull/148/commits/d6646d8a12d97b50f0c79bebd9fb5e02105284ac

Reply:
IMO I don't think it's necessary that we error on both `result` and `error` being present, we can just rely on people following the spec, and if they don't that is their own problem.

It would still be impossible to create JSON that has both members using this crate.

And I don't really want to do a custom impl because of the complexity and I think it' necessary that we allow unknown fields

Reply:
> we can just rely on people following the spec, and if they don't that is their own problem.

I think we should enforce the spec since this is an invariant that checks future code/behavior as well. On the client side, receiving a `Response` that has both `result/error` doesn't make logical sense, assuming and serializing into one of the fields is incorrect behavior.

I.e. this should error:
```rust
// from some server
let j = json!({
    "jsonrpc": "2.0",
    "id": 0,
    "error": e,
    "result": r
});

// client
serde_json::from_value::<Response<String>>(j).unwrap();
```

> I don't really want to do a custom impl because of the complexity and I think it' necessary that we allow unknown fields

I don't think it's _too_ complex. How does a manual impl for only `Response` sound? We enforce the spec, and also allow unknown fields. `Request` can be done with macros since there's nothing to enforce. We would only allow lower-case.

Reply:
Implementation: https://github.com/Cuprate/cuprate/blob/007c9adac735654a9f5bceb06fd13374aed8d729/rpc/json-rpc/src/response.rs#L203-L345

A bit long but I don't think it's too bad.